rfcs/003-nix-nickel.md

```diff
@@ -0,0 +1,460 @@
+computations as data dependencies.
+
+Concretely, a derivation would be represented simply as a recursive record:
+
+```nickel
+builders.derivation = {
+
+  # Interface
+
+  name
+    | doc "Name of the derivation, used in the Nix store path."
+    | Str,
+
+  version
+    | doc "Version of the derivation, used in the Nix store path."
+    | Str,
+    | default = "",
+
+  builder
+    | doc "Command to be executed to build the derivation."
+    | Path,
+
+  args
+    | doc "Arguments passed to the builder."
+    | Array Str
+    | default = [],
+
+  outputs
+    | doc "Symbolic names of the outputs of this derivation."
+    | Array Str
+    | default = ["out"],
+
+  env
+    | doc "Structured values passed to the builder."
+    | {_: Str}
+    # inherit doesn't exist in Nickel, but let's pretend, for conciseness
+    = {inherit outputs},
+
+  # Implementation
+
+  drv
+    | doc "The resulting store derivation."
+    =
+    (env & {
+      name = "%{name}-%{version}";
+      inherit builder args;
+    })
+    |> builtin.derivation
+}
+
+# ... other definitions building on this one
+
+# actual package
+pkgs.hello = builders.unix_package & {
+  name = "hello",
+  version = "1.12",
+  description = "A program that produces a familiar, friendly greeting",
+  license = licenses.gpl,
+
+  enable_gui
+    | doc "Enable GTK+ support."
+    | Bool
+    | default false,
+
+  src = builders.fetchurl & { url = ..., sha256 = ... },
+
+  buildInputs = if enable_gui then [ gtk ] else [],
+}
+```
+
+Thanks to laziness, `nix` could extract fields like `name` or `version` directly
+without having to provide inputs or to evaluate `drv`.
+
+We wall call this approach the **PARM** (**P**ackage **A**s **R**records
+**M**odel) thereafter.
+
+#### Specifying dependencies
+
+The `pkgs.hello` example unveils a difficulty about specifying dependencies. It
+would work in a model where all the packages are defined in one place and
+populate the same huge record that will give Nickelpkgs. But that would be
+totally impracticable and anti-modular.
+
+Writing the `pkgs.hello` in its own file in Nickel today would result in the
+`gtk` occurrence raising an `unbound variable` error, and for a good reason:
+`gtk` is used but seemingly defined nowhere. In practice, `gtk` is expected to
+appear in the final recursive record that would Nickelpkgs. But from a language
+point of view, allowing references to yet-to-be-defined identifiers, that is
+having dynamic scoping, comes with its share of drawbacks: non-locality making
+code harder to understand, inability to detect unbound variables statically
+(incurring late error reporting), inability to properly typecheck such code,
+etc.
+
+Let us review possible suctions:
+
+##### Explicit list
+
+One possibility is to specify explicitly all the
+dependencies as record fields without definition:
+
+```nickel
+# file hello.ncl
+{
+  # dependencies
+  gtk | NickelPackage,
```

Comment:
this could allow inversion of control, where a child package can modify a parent package

```nickel
# file hello.ncl
{
  # dependencies
  gtk | NickelPackage = {
    version = "2.3.4", # bottom-up pinning. nixpkgs: gtk_2_3_4
    enable_somefeature = true, # require feature from parent
    # nixpkgs: hello = callpackage ./hello.nix { gtk = gtk.override { enable_somefeature = true; }; };
  },
```

problem: pollution &rarr; move to scope

```nickel
# file hello.ncl
{
  pkgs = {
    gtk | NickelPackage,
  },
```

Reply:
Good point. Although this could also be solved by merge priorities, the Nickelpkg layer erasing the local definitions. But it does sound fragile. Another issue is that having everything living at the same level forces all the versions of gtk to be the same in the final fixpoint, preventing local overrides (we want `this` package to have `python=python27` but `that` package to have `python=python3` for example).

For those reasons, I think your first suggestion sounds like the best way forward. Also, it could avoid having to repeat dependencies as say `inputs` and `build_inputs`: we could kill two bird with one stone and declare both inputs for Nix and the "merging interface" for Nickel as in

```nickel
{
  inputs = {
    gtk,
  },

 build_inputs = {
    foocompiler,
  }

  # ... inputs.gtk ...
}
```

Reply:
bikeshed:
inputs sounds too general, could be packages and options (enable_somefeature)
packages is too long
pkgs is short & sweet (but also used in nixpkgs)

or we identify "package scopes" by type

```
{
  p | Pkgs = {
    gtk,
  },
  rundeps | Pkgs = [
    apkg, bpkg, cpkg, dpkg,
  ],
  drv = {
    build = m%"
      cp %{p.gtk}/lib/libgtk.so $out/lib/
      echo PATH=%{lib.makePath rundeps} >$out/bin/wrapper.sh
    "%m,
  },
}
```

such short names would compensate for nickel's missing `with` feature